src/container/container_iv.c

```diff
@@ -373,46 +437,165 @@ cont_iv_snapshots_fetch(void *ns, uuid_t cont_uuid, uint64_t **snapshots,
+	if (iv_entry->iv_snap.snap_cnt == 0) {
 		*snap_count = 0;
 		D_GOTO(free, rc = 0);
 	}
 
-	D_ALLOC(*snapshots, sizeof(iv_entry->snaps[0]) * iv_entry->snap_size);
+	D_ALLOC(*snapshots,
+	      sizeof(iv_entry->iv_snap.snaps[0]) * iv_entry->iv_snap.snap_cnt);
 	if (*snapshots == NULL)
 		D_GOTO(free, rc = -DER_NOMEM);
 
-	memcpy(*snapshots, iv_entry->snaps,
-	       sizeof(iv_entry->snaps[0]) * iv_entry->snap_size);
-	*snap_count = iv_entry->snap_size;
+	memcpy(*snapshots, iv_entry->iv_snap.snaps,
+	       sizeof(iv_entry->iv_snap.snaps[0]) * iv_entry->iv_snap.snap_cnt);
+	*snap_count = iv_entry->iv_snap.snap_cnt;
 free:
 	D_FREE(iv_entry);
 	return rc;
 }
 
+struct iv_capa_ult_arg {
+	uuid_t		pool_uuid;
+	uuid_t		cont_uuid;
+	uuid_t		cont_hdl_uuid;
+	ABT_eventual	eventual;
+};
+
+static void
+cont_iv_capa_refresh_ult(void *data)
+{
+	struct iv_capa_ult_arg	*arg = data;
+	struct cont_iv_entry	entry = { 0 };
+	struct ds_pool		*pool;
+	int			rc;
+
+	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+
+	pool = ds_pool_lookup(arg->pool_uuid);
+	if (pool == NULL)
+		D_GOTO(out, rc = -DER_NONEXIST);
+
+	rc = cont_iv_fetch(pool->sp_iv_ns, IV_CONT_CAPA,
+			   arg->cont_hdl_uuid, &entry, sizeof(entry));
+	if (rc)
+		D_GOTO(out, rc);
+
+	uuid_copy(arg->cont_uuid, entry.cont_uuid);
+out:
+	ABT_eventual_set(arg->eventual, (void *)&rc, sizeof(rc));
+}
+
+int
+cont_iv_capa_fetch(uuid_t pool_uuid, uuid_t cont_hdl_uuid,
+		   uuid_t cont_uuid, struct ds_cont_hdl **cont_hdl)
+{
+	struct iv_capa_ult_arg	arg;
+	ABT_eventual		eventual;
+	int			*status;
+	int			rc;
+
+	if (!DAOS_FAIL_CHECK(DAOS_FORCE_CAPA_FETCH)) {
+		*cont_hdl = ds_cont_hdl_lookup(cont_hdl_uuid);
+		if (*cont_hdl != NULL)
+			return 0;
+	}
+
+	D_DEBUG(DB_TRACE, "Can not find "DF_UUID" hdl\n",
+		DP_UUID(cont_hdl_uuid));
+
+	/* Fetch the capability from the leader */
+	rc = ABT_eventual_create(sizeof(*status), &eventual);
+	if (rc != ABT_SUCCESS)
+		return dss_abterr2der(rc);
+
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	uuid_copy(arg.cont_hdl_uuid, cont_hdl_uuid);
+	arg.eventual = eventual;
+	rc = dss_ult_create(cont_iv_capa_refresh_ult, &arg,
+			    DSS_ULT_POOL_SRV, 0, 0, NULL);
+	if (rc)
+		D_GOTO(out_eventual, rc);
+
+	rc = ABT_eventual_wait(eventual, (void **)&status);
+	if (rc != ABT_SUCCESS)
+		D_GOTO(out_eventual, rc = dss_abterr2der(rc));
+	if (*status != 0)
+		D_GOTO(out_eventual, rc = *status);
+
+	if (uuid_compare(arg.cont_uuid, cont_uuid) != 0) {
+		D_ERROR("different cont "DF_UUID" != "DF_UUID"\n",
+			DP_UUID(arg.cont_uuid), DP_UUID(cont_uuid));
+		D_GOTO(out_eventual, rc = -DER_NONEXIST);
+	}
+
+	*cont_hdl = ds_cont_hdl_lookup(cont_hdl_uuid);
+	if (*cont_hdl == NULL) {
+		D_DEBUG(DB_TRACE, "Can not find "DF_UUID" hdl\n",
+			DP_UUID(cont_hdl_uuid));
+		D_GOTO(out_eventual, rc = -DER_NONEXIST);
+	}
+
+out_eventual:
+	ABT_eventual_free(&eventual);
+	return rc;
+}
+
+int
+cont_iv_capability_update(void *ns, uuid_t cont_hdl_uuid, uuid_t cont_uuid,
```

Comment:
inconsistent name right? previous function is cont_iv_capa_, this one is cont_iv_capability_

Reply:
No ULT for this function?

Reply:
oh, for now, this will only happen on xstream 0, so does not need ult for this. but I should add a assertion indeed.